### PR TITLE
[coordinator] Support for augmenting tags in mapping rules

### PR DESCRIPTION
**Pull request by shreyassrivatsan**
**What this PR does / why we need it**:
This change adds the ability to augment additional tags on metrics targeted towards graphite ingestion.
- `__m3_graphite_aggregation__` adds the aggregation tag as the last numbered tag for graphite metrics
- `__m3_graphite_prefix__` prepends the provided path to the appropriate numbered tags.

Secondly, the change additionally adds support for metric types and allows to specify metrics types in mapping/rollup rules.

Thirdly, the change provides a config flag to disable auto mapping rules.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
```release-note
Adds the ability to disable auto mapping rules in downsampler configuration. Additionally adds support for graphite metrics in the remote write endpoint. For graphite, provides the ability to specify a prefix for the metrics and for an aggregation tag to be appended at end of the path.
```

**Does this PR require updating code package or user-facing documentation?**:
Documentation for mapping rules need to be updated.
```documentation-note

```

### CODE REVIEW

File: scripts/docker-integration-tests/prometheus/m3coordinator.yml [hunk 1]
```diff
@@ -41,5 +41,15 @@ clusters:
...
 
+downsample:
+  rules:
+    mappingRules:
+    - filter: __m3_type__:gauge
+      name: aggregate gauges
+      aggregations: ["Last"]
+      storagePolicies:
+      - resolution: 15s
+        retention: 10h
```
Comment (robskillington) [inline scripts/docker-integration-tests/prometheus/m3coordinator.yml:52]: Hm, shouldn't we auto apply this? Why is this required in the integration tests?

It will be backwards incompatible if we require to add this rule in place of the automapping rules here before.

Comment (shreyassrivatsan) [reply scripts/docker-integration-tests/prometheus/m3coordinator.yml:52]: Yeah, shouldn;t need this.. let me remove.. this change was from a previous version of the diff where I had removed auto mapping rules..

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 2]
```diff
@@ -327,6 +339,132 @@ func (a *metricsAppender) Finalize() {
...
+	if err := a.tagEncoder.Encode(tags); err != nil {
+		return samplesAppender{}, err
+	}
+	data, ok := a.tagEncoder.Data()
+	if !ok {
+		return samplesAppender{}, fmt.Errorf("unable to encode tags: names=%v, values=%v", tags.names, tags.values)
+	}
+	// NB (@shreyas): if this is the last use of the tagEncoder we do not need
+	// this copy, but its hard to know and enforce that.
+	id := make([]byte, data.Len())
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:397]: This seems like will add a lot more allocs in the general case, why do we know this is not the last usage?

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 3]
```diff
@@ -171,6 +184,7 @@ func (a *metricsAppender) SamplesAppender(opts SampleAppenderOptions) (SamplesAp
 		// mapping rule has provided an override for a storage policy,
 		// if so then skip aggregating for that storage policy).
 		for idx, stagedMetadatasProto := range a.defaultStagedMetadatasProtos {
+			fmt.Printf("going over staged metadatas %v\n", stagedMetadatasProto)
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:187]: Remove printf?

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go (hunk 2 above)
Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:397]: I guess I can pass in a flag for the last usage.. So that should be the general case.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 4]
```diff
@@ -327,6 +339,132 @@ func (a *metricsAppender) Finalize() {
...
+	stagedMetadata metadata.StagedMetadata,
+) ([]samplesAppender, error) {
+	var (
+		pipelines []metadata.PipelineMetadata
+		appenders []samplesAppender
+	)
+	for _, pipeline := range stagedMetadata.Pipelines {
+		// For pipeline which have tags to augment we generate and send
+		// separate IDs. Other pipelines return the same.
+		pipeline := pipeline
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:352]: Hm, is this needed to capture the var or something?

Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:351]: not needed currently..

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 5]
```diff
@@ -109,6 +117,12 @@ func (a *metricsAppender) SamplesAppender(opts SampleAppenderOptions) (SamplesAp
 	matchResult := a.matcher.ForwardMatch(id, fromNanos, toNanos)
 	id.Close()
 
+	// Filter out the m3 custom tags once the ForwardMatch is complete. If
+	// modified set the unowned ID to nil so that we know to recreate it later.
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:121]: nit: Can we use a bool here instead? Using nil as implicit state seems a little less readable.
i.e. 
```go
needRecomputeID := false

//... 

if modified := a.tags.filterPrefix(m3MetricsPrefix); modified {
  needRecomputeID = true
}

if needRecomputeID {
  unownedID = ....
}
```

Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:164]: let me clean this up.. That's actually not used anywhere and we just compute it before we send stuff to aggregator. Also, the filter will always filter out stuff as we add the metrics type tag all the time..

File: src/cmd/services/m3coordinator/ingest/write.go [hunk 6]
```diff
@@ -437,6 +442,33 @@ func (d *downsamplerAndWriter) writeAggregatedBatch(
...
+			// through the downsampler in general, but right now the aggregator
+			// does not allow context to be attached to a metric so when it calls
+			// back the context is lost currently.
+			// TODO_FIX_GRAPHITE_TAGGING: Using this string constant to track
+			// all places worth fixing this hack. There is at least one
+			// other path where flows back to the coordinator from the aggregator
+			// and this tag is interpreted, eventually need to handle more cleanly.
+			appender.AddTag(downsample.MetricsOptionIDSchemeTagName,
+				downsample.GraphiteIDSchemeTagValue)
+		}
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/ingest/write.go:470]: Let's move this into SamplesAppender as mentioned.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 7]
```diff
@@ -42,6 +45,11 @@ import (
 	"go.uber.org/zap/zapcore"
 )
 
+var (
+	m3MetricsPrefix              = []byte("__m3")
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:49]: As discussed, let's only enable adding type tag and ID re-encoding if and only if rules use special `__m3` tags.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 8]
```diff
@@ -370,6 +378,142 @@ func (a *metricsAppender) Finalize() {
...
+		return samplesAppender{}, err
+	}
+	data, ok := a.tagEncoder.Data()
+	if !ok {
+		return samplesAppender{}, fmt.Errorf("unable to encode tags: names=%v, values=%v", tags.names, tags.values)
+	}
+	// NB (@shreyas): if this is the last use of the tagEncoder we do not need
+	// this copy. The caller tells us what to do here.
+	var id []byte
+	if copyID {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:442]: As discussed, perhaps we can use two slices "cachedTagEncoders" and "inuseTagEncoders" to avoid in the regular case needing to alloc here for "extra" IDs?

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 9]
```diff
@@ -42,6 +46,17 @@ import (
...
 
+var (
+	m3TypeTag      = []byte("__m3_type__")
+	m3CounterValue = []byte("counter")
+	m3GaugeValue   = []byte("gauge")
+	m3TimerValue   = []byte("timer")
+
+	m3MetricsPrefix              = []byte("__m3")
+	m3MetricsPrefixString        = string(m3MetricsPrefix)
+	m3MetricsGraphiteAggregation = []byte("__m3_graphite_aggregation__")
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:57]: nit: Can you make this use a shared M3 prefix variable to avoid them diverging in future? (can accept my suggestion if you think that's ok)
```suggestion
	m3MetricsPrefix       = []byte("__m3")
	m3MetricsPrefixString = string(m3MetricsPrefix)

	m3TypeTag      = []byte(m3MetricsPrefixString + "_type__")
	m3CounterValue = []byte("counter")
	m3GaugeValue   = []byte("gauge")
	m3TimerValue   = []byte("timer")

	m3MetricsGraphiteAggregation = []byte(m3MetricsPrefixString + "_graphite_aggregation__")
```

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 10]
```diff
@@ -103,8 +120,13 @@ func newMetricsAppender(pool *metricsAppenderPool) *metricsAppender {
...
 	a.metricsAppenderOptions = opts
-	if a.tagEncoder == nil {
-		a.tagEncoder = opts.tagEncoderPool.Get()
+
+	// Copy over any previous inuse encoders to the cached encoders list. If
+	// there are no cached encoders then create one.
+	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
+	a.inuseEncoders = a.inuseEncoders[:0]
+	if len(a.cachedEncoders) == 0 {
+		a.cachedEncoders = append(a.cachedEncoders, opts.tagEncoderPool.Get())
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:129]: nit: Do you need to do this? Why not let "one place" do this perhaps, in the call to `a.tagEncoder()` which already lazy allocs the tag encoder?

Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:130]: true, dont need this here. will remove.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 11]
```diff
@@ -103,8 +120,13 @@ func newMetricsAppender(pool *metricsAppenderPool) *metricsAppender {
 // reset is called when pulled from the pool.
 func (a *metricsAppender) reset(opts metricsAppenderOptions) {
 	a.metricsAppenderOptions = opts
-	if a.tagEncoder == nil {
-		a.tagEncoder = opts.tagEncoderPool.Get()
+
+	// Copy over any previous inuse encoders to the cached encoders list. If
+	// there are no cached encoders then create one.
+	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
+	a.inuseEncoders = a.inuseEncoders[:0]
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:128]: 1) Need to nil out any refs before shrinking slice to `inuseEncoders[:0]` otherwise could hang onto it strongly forever.
2) Can we put this in a helper? I see this done other places later in the code.
```go
	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
	for i := range a.inuseEncoders {
		a.inuseEncoders[i] = nil
	}
	a.inuseEncoders = a.inuseEncoders[:0]
```

File: src/cmd/services/m3coordinator/downsample/options.go [hunk 12]
```diff
@@ -648,6 +681,9 @@ func (cfg Configuration) newAggregator(o DownsamplerOptions) (agg, error) {
 		rs := rules.NewEmptyRuleSet(defaultConfigInMemoryNamespace,
 			updateMetadata)
 		for _, mappingRule := range cfg.Rules.MappingRules {
+			if strings.Contains(mappingRule.Filter, m3MetricsPrefixString) {
+				m3PrefixFilter = true
+			}
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/options.go:686]: nit: Want to do this in the rollup rules too? (see if they use the filter at all?)?

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go (hunk 11 above)
Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:128]: done

File: src/cmd/services/m3coordinator/downsample/options.go (hunk 12 above)
Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/options.go:686]: sure, we do not do any specific handling for rollup rules yet, but would be good to do for completeness.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 13]
```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
...
+
+func (a *metricsAppender) tagEncoder() serialize.TagEncoder {
+	// Take an encoder from the cached encoder list, if not present get one
+	// from the pool. Add the returned encoder to the used list.
+	var tagEncoder serialize.TagEncoder
+	if len(a.cachedEncoders) == 0 {
+		tagEncoder = a.tagEncoderPool.Get()
+	} else {
+		tagEncoder = a.cachedEncoders[0]
+		a.cachedEncoders = a.cachedEncoders[1:]
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:428]: Perhaps better to take from the tail, nil out the tail entry index, then shrink so the slice doesn't endlessly grow by move to the right always?

Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:428]: true yeah, that's better.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 14]
```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
...
+}
+
+func (a *metricsAppender) newSamplesAppenders(
+	stagedMetadata metadata.StagedMetadata,
+	unownedID []byte,
+) ([]samplesAppender, error) {
+	// Check if any of the pipelines have tags or a graphite prefix to set.
+	var tagsExist bool
+	for _, pipeline := range stagedMetadata.Pipelines {
+		if len(pipeline.Tags) != 0 || len(pipeline.GraphitePrefix) > 0 {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:442]: nit: Consistency wise, maybe both `> 0` or both `!= 0`?

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 15]
```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
...
+	}
+	a.inuseEncoders = append(a.inuseEncoders, tagEncoder)
+	tagEncoder.Reset()
+	return tagEncoder
+}
+
+func (a *metricsAppender) newSamplesAppenders(
+	stagedMetadata metadata.StagedMetadata,
+	unownedID []byte,
+) ([]samplesAppender, error) {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:438]: This code didn't allocate before, it used to call `a.multiSamplesAppender.addSamplesAppender(samplesAppender{...})` which used a cached slice under the hood in the multiSamplesAppender. Can we make it do the same so we're not allocating a slice here to return something?

Comment (robskillington) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:438]: Can we just make this not return anything and simply call it "addSamplesAppenders" or something and have it call "a.multiSamplesAppender.addSamplesAppender(...)" internally instead of return `[]samplesAppender` which then directly gets added to the multiSamplesAppender anyhow?

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 16]
```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
...
+}
+
+func augmentTags(
+	originalTags *tags,
+	graphitePrefix [][]byte,
+	t []models.Tag,
+	id aggregation.ID,
+) *tags {
+	// Create the prefix tags if any.
+	tags := newTags()
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:526]: This seems like another "allocation per thing processed", can we use the same pattern here with "cachedEncoders" and "inuseEncoders" to avoid that? It's fine to call "newTags()" I think but amortizing when that happens so that in the steady state it doesn't need to do that would be ideal.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go (hunk 15 above)
Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:438]: yeah, can just make that method add to the multi samples appender.. will call the method `addSamplesAppenders`

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go (hunk 16 above)
Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:526]: sure, yeah.. that should help a bit at least..

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go (hunk 15 above)
Comment (shreyassrivatsan) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:438]: didnt see the comment above, but yeah did that.

File: src/cmd/services/m3coordinator/downsample/samples_appender.go [hunk 17]
```diff
@@ -137,6 +137,10 @@ func (a *multiSamplesAppender) addSamplesAppender(v samplesAppender) {
 	a.appenders = append(a.appenders, v)
 }
 
+func (a *multiSamplesAppender) addSamplesAppenders(v []samplesAppender) {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/samples_appender.go:140]: Can we remove this method so callers don't need to allocate a slice? Just call it multiple times?

File: src/cmd/services/m3coordinator/ingest/write.go [hunk 18]
```diff
@@ -446,6 +461,7 @@ func (d *downsamplerAndWriter) writeAggregatedBatch(
 				},
 			}
 		}
+		opts.MetricType = value.Attributes.Type
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/ingest/write.go:464]: nit: Instead of setting this here can you set it in the original declaration?
```go
opts := downsample.SampleAppenderOptions{
  MetricsType: value.Attributes.Type,
}
if downsampleMappingRuleOverrides, ok := d.downsampleOverrideRules(overrides); ok {
  // ...
}
```

File: src/metrics/rules/mapping.go [hunk 19]
```diff
@@ -47,6 +50,9 @@ var (
 	errMappingRuleSnapshotIndexOutOfRange                  = errors.New("mapping rule snapshot index out of range")
 	errNilMappingRuleSnapshotProto                         = errors.New("nil mapping rule snapshot proto")
 	errNilMappingRuleProto                                 = errors.New("nil mapping rule proto")
+
+	m3MetricsGraphitePrefix = []byte("__m3_graphite_prefix__")
```
Comment (robskillington) [inline src/metrics/rules/mapping.go:54]: Can you reference the declaration in the other package so we have this string declared in one place? Either that or move the string declaration to here and reference it from the other package (if the imports don't worry the first way).

Comment (shreyassrivatsan) [reply src/metrics/rules/mapping.go:54]: ah yes. its not exported there. let me see where best to put this..

File: src/cmd/services/m3coordinator/downsample/options.go [hunk 20]
```diff
@@ -24,8 +24,11 @@ import (
 	"errors"
 	"fmt"
 	"runtime"
+	"strings"
 	"time"
 
+	"github.com/m3db/m3/src/metrics/metric"
+
```
Comment (shreyassrivatsan) [inline src/cmd/services/m3coordinator/downsample/options.go:31]: will fix.

Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/options.go:31]: nit: No endline here? All first party imports.
